components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/impl/ApisApiServiceImpl.java

```diff
@@ -470,22 +470,48 @@ public Response apisApiIdDocumentsPost(String apiId, DocumentDTO body, String co
         try {
             APIProvider apiProvider = RestApiUtil.getLoggedInUserProvider();
             Documentation documentation = DocumentationMappingUtil.fromDTOtoDocumentation(body);
+            String documentName = body.getName();
             String tenantDomain = RestApiUtil.getLoggedInUserTenantDomain();
 
+            if (body.getType() == DocumentDTO.TypeEnum.OTHER && StringUtils.isBlank(body.getOtherTypeName())) {
+                //check otherTypeName for not null if doc type is OTHER
+                throw RestApiUtil.buildBadRequestException("otherTypeName cannot be empty if type is OTHER.");
+            }
+
+            String sourceUrl = body.getSourceUrl();
+            if (body.getSourceType() == DocumentDTO.SourceTypeEnum.URL &&
+                    (StringUtils.isBlank(sourceUrl) || !RestApiUtil.isURL(sourceUrl))) {
+                throw RestApiUtil.buildBadRequestException("Invalid document sourceUrl Format");
+            }
+
             //this will fail if user does not have access to the API or the API does not exist
             APIIdentifier apiIdentifier = APIMappingUtil.getAPIIdentifierFromApiIdOrUUID(apiId, tenantDomain);
 
+            if (apiProvider.isDocumentationExist(apiIdentifier, documentName)) {
+                String errorMessage = "Requested document '" + documentName + "' already exists";
+                throw RestApiUtil.buildConflictException(errorMessage);
+            }
+
             apiProvider.addDocumentation(apiIdentifier, documentation);
             String newDocumentId = documentation.getId();
 
             //retrieve the newly added document
             documentation = apiProvider.getDocumentation(newDocumentId, tenantDomain);
             DocumentDTO newDocumentDTO = DocumentationMappingUtil.fromDocumentationToDTO(documentation);
-            return Response.status(Response.Status.CREATED).header("Location",
-                    "/apis/" + apiId + "/documents/" + documentation.getId()).entity(newDocumentDTO).build();
+
+            String uriString = RestApiConstants.RESOURCE_PATH_DOCUMENTS_DOCUMENT_ID
+                    .replace(RestApiConstants.APIID_PARAM, apiId)
+                    .replace(RestApiConstants.DOCUMENTID_PARAM, newDocumentId);
+            URI uri = new URI(uriString);
+            return Response.created(uri).entity(newDocumentDTO).build();
         } catch (APIManagementException e) {
-            throw new InternalServerErrorException(e);
+            String errorMessage = "Error while adding the document for API : " + apiId;
+            handleException(errorMessage, e);
```

Comment:
handleException throws an APIManagementException right? If so we're catching and throwing the same exception here. If there's no valid reason to do that (ex: To change the error message) we should remove this catch block.